clients/drcachesim/tracer/kcore_copy.cpp

```diff
@@ -0,0 +1,433 @@
+    if (to_kallsyms_fd < 0) {
+        ASSERT(false, "failed to open kallsyms file");
+        dr_close_file(from_kallsyms_fd);
+        return false;
+    }
+
+    char buf[1024];
+    ssize_t bytes_read;
+    while ((bytes_read = dr_read_file(from_kallsyms_fd, buf, sizeof(buf))) > 0) {
+        if (dr_write_file(to_kallsyms_fd, buf, bytes_read) != bytes_read) {
+            ASSERT(false, "failed to write kallsyms file");
+            dr_close_file(to_kallsyms_fd);
+            dr_close_file(from_kallsyms_fd);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool
+kcore_copy_t::read_modules()
+{
+    std::ifstream f(MODULES_FILE_PATH, std::ios::in);
+    if (!f.is_open()) {
+        ASSERT(false, "failed to open " MODULES_FILE_PATH);
+        return false;
+    }
+    proc_module_t *last_module = modules_;
+    std::string line;
+    while (std::getline(f, line)) {
+        /* Each line is similar to the following line:
+         * 'scsi_dh_hp_sw 12895 0 - Live 0xffffffffa005e000'
+         * We parse the second and the last field to construct a proc_module_t type
+         * instance.
+         */
+        uint64_t addr = 0x0;
+        int len = 0;
+        if (dr_sscanf(line.c_str(), "%*s %d %*d %*s %*s " HEX64_FORMAT_STRING, &len,
+                      &addr) != 2) {
+            ASSERT(false, "failed to parse " MODULES_FILE_PATH);
+            f.close();
+            return false;
+        }
+
+        proc_module_t *module = (proc_module_t *)dr_global_alloc(sizeof(proc_module_t));
+        module->start = addr;
+        module->end = addr + len;
+        module->next = nullptr;
+        kcore_code_segments_num_++;
+        if (last_module == nullptr) {
+            modules_ = module;
+            last_module = module;
+        } else {
+            last_module->next = module;
+            last_module = module;
+        }
+    }
+    f.close();
+    return true;
+}
+
+bool
+kcore_copy_t::read_kallsyms()
+{
+    std::ifstream f(KALLSYMS_FILE_PATH, std::ios::in);
+    if (!f.is_open()) {
+        ASSERT(false, "failed to open " KALLSYMS_FILE_PATH);
+        return false;
+    }
+    proc_module_t *kernel_module = nullptr;
+    std::string line;
+    while (std::getline(f, line)) {
+        char name[KERNEL_SYMBOL_MAX_LEN];
+        uint64_t addr;
+        if (dr_sscanf(line.c_str(), HEX64_FORMAT_STRING " %*1c %299s [%*99s", &addr,
+                      name) < 2)
+            continue;
+        if (strcmp(name, "_stext") == 0) {
+            if (kernel_module != nullptr) {
+                f.close();
+                return false;
+            }
+            kernel_module = (proc_module_t *)dr_global_alloc(sizeof(proc_module_t));
+            kernel_module->start = addr;
+        } else if (strcmp(name, "_etext") == 0) {
+            if (kernel_module == nullptr) {
+                f.close();
+                return false;
+            }
+            kernel_module->end = addr;
```

Comment:
assert that kernel_module is not nullptr

Reply:
Done.